### ISSUE TITLE
Invariant Violation when rendering using both ReactDOM and ReactDOMServer

### ISSUE THREAD

**Issue by BS-Harou** [facebook/react#8487]
Using both ReactDOM.render and ReactDOMServer.render(\w+) causes an invariant violation in React 15.4.1.

See console on https://jsfiddle.net/martinkadlec/jb7bqg5o/4/
There is:
```javascript
react-dom.js:18121 Warning: Exception thrown by hook while handling onSetChildren: Invariant Violation: Item must have been set
Invariant Violation: Item must have been set
    at invariant (https://unpkg.com/react@15.4.1/dist/react-with-addons.js:5175:15)
```
Notice that I render completely different components with each of the render methods.

Some criteria have to be met for the warning to show, like at least two nested components.

To me it seems like the ReactDOMServer is messing with itemMap in the regular React and the IDs of components doesn't match.

**Comment by Aweary** [comment]
Thanks for the report @BS-Harou. It looks like its related to `ReactDOM` and `ReactDOMServer` are using `ReactComponentTreeHook` which uses [a single Map](https://github.com/facebook/react/blob/32f5b034ed229d048f76ae74e18d270edc801dbf/src/isomorphic/hooks/ReactComponentTreeHook.js#L75-L76) to store items, so when when one deletes an item it may throw this error if the other tries accessing it.

**Comment by gaearon** [comment]
To be clear it won't crash your code since the error is being caught. But we need to fix it.

**Comment by captbaritone** [comment]
I have a small reproducible example of this bug. I can create an example repo that triggers it if needed.

**Comment by BS-Harou** [comment]
It seems my fiddle stop working for some reason. @captbaritone go ahead :)

**Comment by mrgurdeep** [comment]
@BS-Harou is there any reason you are using ReactDOM.render() twice with same values. I removed one and it solved the warning was removed.

**Comment by BS-Harou** [comment]
@mrgurdeep There is. The internal itemMap object gets "damaged" only after both React and ReactDOMServer render methods are used. So only the next render method called after that tries to use the "damaged" itemMap and triggers the warning. 

Note that I don't think it matters whether the second ReactDOM.render is called with the same data as the first one or not.

**Comment by gaearon** [comment]
Looks like we have a fix in #9005—going to merge it once it's rebased, and hopefully get it into 15.5.

**Comment by gaearon** [comment]
It was fixed in 15.5.
